// @lingo-reader/shared/dist/index.js
'use strict';

var path$1 = require('node:path');
var events = require('events');
var sax = require('sax');

function resolve(...args) {
  return path$1.resolve(...args);
}
function basename(p, ext) {
  return path$1.basename(p, ext);
}
function extname(p) {
  return path$1.extname(p);
}
function dirname(p) {
  return path$1.dirname(p);
}
function isAbsolutePosix(p) {
  return path$1.posix.isAbsolute(p);
}
function joinPosix(...paths) {
  return path$1.posix.join(...paths);
}
var path = {
  resolve,
  basename,
  extname,
  dirname,
  isAbsolutePosix,
  joinPosix
};

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
function stripBOM(str) {
  if (str.charCodeAt(0) === 65279) {
    return str.slice(1);
  }
  return str;
}
const defaults = {
  0.1: {
    explicitCharkey: false,
    trim: true,
    // normalize implicates trimming, just so you know
    normalize: true,
    // normalize tag names to lower case
    normalizeTags: false,
    // set default attribute object key
    attrkey: "@",
    // set default char object key
    charkey: "#",
    // always put child nodes in an array
    explicitArray: false,
    // ignore all attributes regardless
    ignoreAttrs: false,
    // merge attributes and child elements onto parent object.  this may cause collisions.
    mergeAttrs: false,
    explicitRoot: false,
    validator: null,
    xmlns: false,
    // fold children elements into dedicated property (works only in 0.2)
    explicitChildren: false,
    childkey: "@@",
    charsAsChildren: false,
    // include white-space only text nodes
    includeWhiteChars: false,
    // callbacks are async? not in 0.1 mode
    async: false,
    strict: true,
    attrNameProcessors: null,
    attrValueProcessors: null,
    tagNameProcessors: null,
    valueProcessors: null,
    emptyTag: ""
  },
  0.2: {
    explicitCharkey: false,
    trim: false,
    normalize: false,
    normalizeTags: false,
    attrkey: "$",
    charkey: "_",
    explicitArray: true,
    ignoreAttrs: false,
    mergeAttrs: false,
    explicitRoot: true,
    validator: null,
    xmlns: false,
    explicitChildren: false,
    preserveChildrenOrder: false,
    childkey: "$$",
    charsAsChildren: false,
    // include white-space only text nodes
    includeWhiteChars: false,
    // not async in 0.2 mode either
    async: false,
    strict: true,
    attrNameProcessors: null,
    attrValueProcessors: null,
    tagNameProcessors: null,
    valueProcessors: null,
    // xml building options
    rootName: "root",
    xmldec: { version: "1.0", encoding: "UTF-8", standalone: true },
    doctype: null,
    renderOpts: { pretty: true, indent: "  ", newline: "\n" },
    headless: false,
    chunkSize: 1e4,
    emptyTag: "",
    cdata: false
  }
};
function normalize(str) {
  return str.toLowerCase();
}
function isEmpty(thing) {
  return typeof thing === "object" && thing !== null && Object.keys(thing).length === 0;
}
function processItem(processors, item, key) {
  for (const process of processors) {
    item = process(item, key);
  }
  return item;
}
function defineProperty(obj, key, value) {
  const descriptor = /* @__PURE__ */ Object.create(null);
  descriptor.value = value;
  descriptor.writable = true;
  descriptor.enumerable = true;
  descriptor.configurable = true;
  Object.defineProperty(obj, key, descriptor);
}
class Parser extends events.EventEmitter {
  constructor(opts) {
    super();
    __publicField(this, "options");
    __publicField(this, "remaining", "");
    __publicField(this, "saxParser");
    __publicField(this, "resultObject", null);
    __publicField(this, "EXPLICIT_CHARKEY");
    // saxParser status
    __publicField(this, "errThrown", false);
    __publicField(this, "ended", false);
    __publicField(this, "processAsync", () => {
      try {
        if (this.remaining.length <= this.options.chunkSize) {
          const chunk = this.remaining;
          this.remaining = "";
          this.saxParser = this.saxParser.write(chunk);
          this.saxParser.close();
        } else {
          const chunk = this.remaining.slice(0, this.options.chunkSize);
          this.remaining = this.remaining.slice(this.options.chunkSize);
          this.saxParser = this.saxParser.write(chunk);
          setTimeout(this.processAsync, 0);
        }
      } catch (err) {
        if (!this.errThrown) {
          this.errThrown = true;
          this.emit("error", err);
        }
      }
    });
    __publicField(this, "assignOrPush", (obj, key, newValue) => {
      if (!(key in obj)) {
        if (!this.options.explicitArray) {
          defineProperty(obj, key, newValue);
        } else {
          defineProperty(obj, key, [newValue]);
        }
      } else {
        if (!Array.isArray(obj[key])) {
          defineProperty(obj, key, [obj[key]]);
        }
        obj[key].push(newValue);
      }
    });
    __publicField(this, "reset", () => {
      this.removeAllListeners();
      this.saxParser = sax.parser(this.options.strict, {
        trim: false,
        normalize: false,
        xmlns: this.options.xmlns
      });
      this.errThrown = false;
      this.saxParser.onerror = (error) => {
        this.saxParser.resume();
        if (!this.errThrown) {
          this.errThrown = true;
          this.emit("error", error);
        }
      };
      this.saxParser.onend = () => {
        if (!this.ended) {
          this.ended = true;
          this.emit("end", this.resultObject);
        }
      };
      this.ended = false;
      this.EXPLICIT_CHARKEY = this.options.explicitCharkey;
      this.resultObject = null;
      const stack = [];
      const attrkey = this.options.attrkey;
      const charkey = this.options.charkey;
      this.saxParser.onopentag = (node) => {
        const obj = {};
        obj[charkey] = "";
        if (!this.options.ignoreAttrs) {
          Object.keys(node.attributes).forEach((key) => {
            if (!(attrkey in obj) && !this.options.mergeAttrs) {
              obj[attrkey] = {};
            }
            const newValue = this.options.attrValueProcessors ? processItem(this.options.attrValueProcessors, node.attributes[key], key) : node.attributes[key];
            const processedKey = this.options.attrNameProcessors ? processItem(this.options.attrNameProcessors, key) : key;
            if (this.options.mergeAttrs) {
              this.assignOrPush(obj, processedKey, newValue);
            } else {
              defineProperty(obj[attrkey], processedKey, newValue);
            }
          });
        }
        obj["#name"] = this.options.tagNameProcessors ? processItem(this.options.tagNameProcessors, node.name) : node.name;
        if (this.options.xmlns) {
          obj[this.options.xmlnskey] = { uri: node.uri, local: node.local };
        }
        stack.push(obj);
      };
      this.saxParser.onclosetag = () => {
        let obj = stack.pop();
        const nodeName = obj["#name"];
        if (!this.options.explicitChildren || !this.options.preserveChildrenOrder) {
          delete obj["#name"];
        }
        let cdata;
        if (obj.cdata === true) {
          cdata = obj.cdata;
          delete obj.cdata;
        }
        const s = stack[stack.length - 1];
        let emptyStr = "";
        if (obj[charkey].match(/^\s*$/) && !cdata) {
          emptyStr = obj[charkey];
          delete obj[charkey];
        } else {
          if (this.options.trim) {
            obj[charkey] = obj[charkey].trim();
          }
          if (this.options.normalize) {
            obj[charkey] = obj[charkey].replace(/\s{2,}/g, " ").trim();
          }
          obj[charkey] = this.options.valueProcessors ? processItem(this.options.valueProcessors, obj[charkey], nodeName) : obj[charkey];
          if (Object.keys(obj).length === 1 && charkey in obj && !this.EXPLICIT_CHARKEY) {
            obj = obj[charkey];
          }
        }
        if (isEmpty(obj)) {
          if (typeof this.options.emptyTag === "function") {
            obj = this.options.emptyTag();
          } else {
            obj = this.options.emptyTag !== "" ? this.options.emptyTag : emptyStr;
          }
        }
        if (this.options.validator) {
          const xpath = `/${stack.map((node) => node["#name"]).concat(nodeName).join("/")}`;
          (() => {
            try {
              obj = this.options.validator(xpath, s && s[nodeName], obj);
            } catch (err) {
              this.emit("error", err);
            }
          })();
        }
        if (this.options.explicitChildren && !this.options.mergeAttrs && typeof obj === "object") {
          if (!this.options.preserveChildrenOrder) {
            const node = {};
            if (this.options.attrkey in obj) {
              node[this.options.attrkey] = obj[this.options.attrkey];
              delete obj[this.options.attrkey];
            }
            if (!this.options.charsAsChildren && this.options.charkey in obj) {
              node[this.options.charkey] = obj[this.options.charkey];
              delete obj[this.options.charkey];
            }
            if (Object.getOwnPropertyNames(obj).length > 0) {
              node[this.options.childkey] = obj;
            }
            obj = node;
          } else if (s) {
            s[this.options.childkey] = s[this.options.childkey] || [];
            const objClone = {};
            Object.keys(obj).forEach((key) => {
              defineProperty(objClone, key, obj[key]);
            });
            s[this.options.childkey].push(objClone);
            delete obj["#name"];
            if (Object.keys(obj).length === 1 && charkey in obj && !this.EXPLICIT_CHARKEY) {
              obj = obj[charkey];
            }
          }
        }
        if (stack.length > 0) {
          this.assignOrPush(s, nodeName, obj);
        } else {
          if (this.options.explicitRoot) {
            const old = obj;
            const newObj = {};
            defineProperty(newObj, nodeName, old);
            obj = newObj;
          }
          this.resultObject = obj;
          this.ended = true;
          this.emit("end", this.resultObject);
        }
      };
      const ontext = (text) => {
        const s = stack[stack.length - 1];
        if (s) {
          s[charkey] += text;
          if (this.options.explicitChildren && this.options.preserveChildrenOrder && this.options.charsAsChildren && (this.options.includeWhiteChars || text.replace(/\\n/g, "").trim() !== "")) {
            s[this.options.childkey] = s[this.options.childkey] || [];
            const charChild = {
              "#name": "__text__"
            };
            charChild[charkey] = text;
            if (this.options.normalize) {
              charChild[charkey] = charChild[charkey].replace(/\s{2,}/g, " ").trim();
            }
            s[this.options.childkey].push(charChild);
          }
        }
        return s;
      };
      this.saxParser.ontext = ontext;
      this.saxParser.oncdata = (text) => {
        const s = ontext(text);
        if (s) {
          s.cdata = true;
        }
      };
    });
    __publicField(this, "parseString", (str, cb) => {
      if (cb && typeof cb === "function") {
        this.on("end", (result) => {
          this.reset();
          cb(null, result);
        });
        this.on("error", (err) => {
          this.reset();
          cb(err);
        });
      }
      try {
        str = str.toString();
        if (str.trim() === "") {
          this.emit("end", null);
          return true;
        }
        str = stripBOM(str);
        if (this.options.async) {
          this.remaining = str;
          setTimeout(this.processAsync, 0);
          return this.saxParser;
        }
        return this.saxParser.write(str).close();
      } catch (err) {
        if (!this.errThrown && !this.ended) {
          this.emit("error", err);
          this.errThrown = true;
        } else if (this.ended) {
          throw err;
        }
      }
    });
    __publicField(this, "parseStringPromise", (str) => {
      return new Promise((resolve, reject) => {
        this.parseString(str, (err, value) => {
          if (err) {
            reject(err);
          } else {
            resolve(value);
          }
        });
      });
    });
    this.options = {};
    Object.keys(defaults["0.2"]).forEach((key) => {
      this.options[key] = defaults["0.2"][key];
    });
    if (opts) {
      Object.keys(opts).forEach((key) => {
        this.options[key] = opts[key];
      });
    }
    if (this.options.xmlns) {
      this.options.xmlnskey = `${this.options.attrkey}ns`;
    }
    if (this.options.normalizeTags) {
      if (!this.options.tagNameProcessors) {
        this.options.tagNameProcessors = [];
      }
      this.options.tagNameProcessors.unshift(normalize);
    }
    this.reset();
  }
}
function parseStringPromise(str, a) {
  let options = {};
  if (typeof a === "object") {
    options = a;
  }
  const parser = new Parser(options);
  return parser.parseStringPromise(str);
}

async function parsexml(str, optionsParserOptions = {}) {
  const result = await parseStringPromise(str, optionsParserOptions);
  return result;
}
function camelCase(str) {
  return str.replace(/-([a-z])/gi, (g) => g[1].toUpperCase());
}

exports.camelCase = camelCase;
exports.parsexml = parsexml;
exports.path = path;
